gui/integrity-dialog.cpp

```diff
@@ -0,0 +1,425 @@
+		_cancelButton->setVisible(false);
+		_cancelButton->setLabel(_("Cancel"));
+		_cancelButton->setCmd(kCleanupCmd);
+
+		break;
+	}
+	}
+}
+
+void IntegrityDialog::handleCommand(CommandSender *sender, uint32 cmd, uint32 data) {
+	switch (cmd) {
+	case kCleanupCmd: {
+		delete g_state;
+		g_state = nullptr;
+
+		close();
+		break;
+	}
+	default:
+		Dialog::handleCommand(sender, cmd, data);
+	}
+}
+
+void IntegrityDialog::handleTickle() {
+	if (_close) {
+		close();
+		_close = false;
+		return;
+	}
+
+	int32 progress = getDownloadingProgress();
+	if (_progressBar->getValue() != progress) {
+		refreshWidgets();
+		g_gui.scheduleTopDialogRedraw();
+	}
+
+	Dialog::handleTickle();
+}
+
+void IntegrityDialog::reflowLayout() {
+	Dialog::reflowLayout();
+	refreshWidgets();
+}
+
+void IntegrityDialog::refreshWidgets() {
+	uint32 progress = getDownloadingProgress();
+	_percentLabel->setLabel(Common::String::format("%u %%", progress));
+	_progressBar->setValue(progress);
+}
+
+void IntegrityDialog::setError(Common::U32String &msg) {
+	_errorText->setLabel(msg);
+
+	_cancelButton->setLabel(_("Close"));
+	_cancelButton->setCmd(kCleanupCmd);
+}
+
+void IntegrityDialog::calculateTotalSize(Common::String gamePath) {
+	const Common::FSNode dir(gamePath);
+
+	if (!dir.exists() || !dir.isDirectory())
+		return;
+
+	Common::FSList fileList;
+	if (!dir.getChildren(fileList, Common::FSNode::kListAll))
+		return;
+
+	if (fileList.empty())
+		return;
+
+	// Process the files and subdirectories in the current directory recursively
+	for (Common::FSList::const_iterator it = fileList.begin(); it != fileList.end(); it++) {
+		const Common::FSNode &entry = *it;
+
+		if (entry.isDirectory())
+			calculateTotalSize(entry.getPath());
+		else {
+			Common::File file;
+			if (!file.open(entry))
+				continue;
+
+			g_state->totalSize += file.size();
+		}
+	}
+}
+
+Common::Array<Common::StringArray> IntegrityDialog::generateChecksums(Common::String gamePath, Common::Array<Common::StringArray> &fileChecksums) {
+	const Common::FSNode dir(gamePath);
+
+	if (!dir.exists() || !dir.isDirectory())
+		return {};
+
+	Common::FSList fileList;
+	if (!dir.getChildren(fileList, Common::FSNode::kListAll))
+		return {};
+
+	if (fileList.empty())
+		return {};
+
+	// Process the files and subdirectories in the current directory recursively
+	for (Common::FSList::const_iterator it = fileList.begin(); it != fileList.end(); it++) {
+		const Common::FSNode &entry = *it;
+
+		if (entry.isDirectory())
+			generateChecksums(entry.getPath(), fileChecksums);
+		else {
+			const Common::Path filename(entry.getPath());
+			auto macFile = Common::MacResManager();
+			if (macFile.openFileOrDataFork(filename)) {
+				debug("got here lol");
+				auto originalStream = macFile.openFileOrDataFork(filename);
+				auto fileStream = originalStream;
+
+				Common::Array<Common::String> fileChecksum = {filename.toString()};
+				// Various checksizes
+				for (auto size : {0, 5000, 1024 * 1024}) {
```

Comment:
for mac files, you need to compute:  r5000, m5000, d5000, rt5000, mt5000, dt5000 and r, d (full). No need to compute 1024 * 1024.